skops/io/_utils.py

```diff
@@ -171,3 +171,21 @@ def get_state(value, save_context):
 
     res["__id__"] = __id__
     return res
+
+
+def get_type_name(
+    t: Any,
+) -> str:
+    """Simple helper function to take in a type, and return its name as a string"""
+    return f"{t.__module__}.{t.__name__}"
+
+
+def cast_defaults_to_strings(defaults: List[Any]) -> List[str]:
+    """Simple helper function that takes in a list of defaults,
+    and converts any types found to strings"""
+    if not defaults:
+        return []
+    if not type(defaults) in [list, tuple]:
+        defaults = [defaults]
+
+    return [get_type_name(t) if type(t) is not str else t for t in defaults]
```

Comment:
Would it not be better to check `if not isinstance(t, str)` or am I missing something?

Reply:
That also works, only difference would be if something was a subclass of `str`, which is unlikely to ever be a problem for us (unless someone wanted to check the type safety of something that extends `str`).

Happy to rework this to be an `isinstance` check instead, I felt it would be safer to explicitly check that something is exactly `str` though

Reply:
Hmm, maybe having a strict check here is more secure? But I can't really come up with a possible scenario where it would matter. FWIW, `UserString` would not pass the `isinstance` check. @adrinjalali any opinion?

Reply:
> Here we only want to compare strings ... So I'd do isinstance here rather than strict type checking

Do I misunderstand or is this a contradiction? If we only want to compare strings and not any subclasses, should we not do the strict check, i.e.  `type(t)`?

So if a user creates a subclass from string and then says `trusted = [MyString("foo.bar")]`, would we trust `foo.bar` or `mytypes.MyString` (or whatever its module is)? (Note that the latter would actually fail right now with `'MyString' object has no attribute '__name__'`)

Reply:
Yeah, that exact case was the reason I did this strict check in the first place, but I think that actually hits on the whole problem.

`MyString("foo.bar")` is no longer a type, it's an instance. So we should in that case treat it as `foo.bar` instead of `MyString`. 

I'll add a test that covers this and change it to support this.

Essentially, the functionality should (I think) be:

`UserDefinedString -> ["module.UserDefinedString"]`
`UserDefinedString("foo") ->["foo"]`

Reply:
I'm not sure what the issue here is, I might be missing something. This is how I think about it:

```python
def get_type_paths(trusted):
	def _get_str(value):
		if isinstance(value, str):
			return value
		if type(value) is type:
			return get_module(value)
		raise Exception()

	return [_get_str(value) for value in trusted]
```

The user either passes types, or instances, for instances we keep them if they're strings (or subclasses), for types, we return their address.

Reply:
> The user either passes types, or instances, for instances we keep them if they're strings (or subclasses), for types, we return their address.

In general, I agree (and that's also how I understand @E-Aho's suggestion), this would be a reasonable approach that's also easy to remember. I just have two concerns:

1. Is there any security issue with allowing custom types there? As I said, I can't come up with any.
2. `collections.UserString` is not an instance of str, if we want to allow user defined strings, should we allow that?

Reply:
I think of supporting that if at some point somebody opens an issue with this request. I don't think it's used enough for us to complicate things for it at this point.

Reply:
Okay, so to summarize, the agreement is to check `isinstance` of `str`, not `type` `str`, not to support `UserString`, and that we're not aware of any security issue caused by `isinstance` vs `type` check.